proto/rdp/README.md

```diff
@@ -0,0 +1,286 @@
+---
+title: Reviewdog Diagnostic Protocol (RDP)
+date: 2020-06-15
+author: haya14busa
+status: Proposed / Experimental
+---
+
+# Status
+
+This document proposes Reviewdog Diagnostic Protocol (RDP) and it's still
+in experimental stage.
+
+Any review, suggestion, feedback, criticism, and comments from anyone is very
+much welcome. Please leave comments in Pull Request ([#629](https://github.com/reviewdog/reviewdog/pull/629))
+, in issue [#628](https://github.com/reviewdog/reviewdog/issues/628) or
+file [an issue](https://github.com/reviewdog/reviewdog/issues).
+
+The document and the actual definition are currently in under
+https://github.com/reviewdog/reviewdog repository, but we may create a separate
+repository once it's reviewed and stabilized.
+
+# Reviewdog Diagnostic Protocol (RDP)
+
+Reviewdog Diagnostic Protocol (RDP. Or **R**eview**D**og **P**rotocol in short.)
+defines standard machine-readable message structures which represents a
+result of diagnostic tool such as a compiler or a linter.
+
+The idea behind the Reviewdog Diagnostic Protocol (RDP) is to standardize
+the protocol for how diagnostic tools (e.g. compilers, linters, etc..) and
+development tools (e.g. editors, reviewdog, code review API etc..) communicate.
+
+See [reviewdog.proto](reviewdog.proto) for the actual definition.
+
+## Wire formats of Reviewdog Diagnostic Protocol.
+
+RDP uses [Protocol Buffer](https://developers.google.com/protocol-buffers) to
+define the message structure, but recommended wire format is JSON considering
+it's widely used and easy to support both from diagnostic tools and development
+tools.
+
+### **rdpjsonl**
+JSON Lines (http://jsonlines.org/) of the [`Diagnostic`](reviewdog.proto) message.
+
+Example:
+```json
+{"message": "<msg>", "location": {"path": "<file path>", "range": {"start": {"line": 14, "column": 15}}}, "severity": "ERROR"}
+{"message": "<msg>", "location": {"path": "<file path>", "range": {"start": {"line": 14, "column": 15}, "end": {"line": 14, "column": 18}}}, "suggestions": [{"range": {"start": {"line": 14, "column": 15}, "end": {"line": 14, "column": 18}}, "text": "<replacement text>"}], "severity": "WARNING"}
+...
+```
+
+### **rdpjson**
+JSON format of the [`DiagnosticResult`](reviewdog.proto) message.
+
+Example:
+```json
+{
+  "source": {
+    "name": "super lint",
+    "url": "https://example.com/url/to/super-lint"
+  },
+  "severity": "WARNING",
+  "diagnostics": [
+    {
+      "message": "<msg>",
+      "location": {
+        "path": "<file path>",
+        "range": {
+          "start": {
+            "line": 14,
+            "column": 15
```

Comment:
Some compilers/linters can only provide absolute position in the file, not line/column.

Reply:
Yeah, that's actually a good discussion point. The reason why I didn't include the byte offset (absolute position in the file) is that if we added it, all consumers tools of RDP output (such as reviewdog) need to add support to convert byte offset <-> line/column. 

My current idea is the conversion should happen outside RDP, such as in linters or converters.